### ISSUE TITLE
Segmentation fault building with flto

### ISSUE THREAD

**Issue by spsaxena** [emscripten-core/emscripten#11432]
In a specific case creating static const& (see example) and building with **flto** leads to runtime segmentation fault 
emscripten : 1.39.10

Example to demonstrate the same
https://github.com/spsaxena/emsc_issues/tree/master/segFault
Output
![Output](https://user-images.githubusercontent.com/12669021/84866052-c4732c80-b0ab-11ea-83f0-1f6aa48fcce7.PNG)

**Comment by sbc100** [comment]
You could try building with `-fsantitize-address` to debug the crash.

**Comment by spsaxena** [comment]
@sbc100 I could not manage to debug using ASan.

If I add Asan, I get the following error -
shared:ERROR: ASan does not work with SAFE_HEAP

If I keep Asan and remove SAFE_HEAP I get a different error -
RuntimeError: abort(RuntimeError: abort(external global '__global_base' is missing. perhaps a side module was not linked in? if this function was expected to arrive from a system library, try to build the MAIN_MODULE with EMCC_FORCE_STDLIBS=1 in the environment)

but, even if I just have 'int main() { return 0; }' and build with those build settings including Asan, I get the same '__global_base' is missing' error. 
So I doubt if ASan supports dyn linking, as I am building with -s MAIN_MODULE=2.

Anyway, I could workaround the problem by changing the code (commenting static), however did not root cause further if it is compiler issue, settings issue or code problem.

**Comment by sbc100** [comment]
Can it work find without MAIN_MODULE / SIDE_MODULE?     The more you can reduce the number of options and reduce the size of your repro case the more likely the root cause will emerge.

**Comment by sbc100** [comment]
Also how about with MAIN_MODULE=1 rather than 2?

**Comment by sbc100** [comment]
I'm trying to reproduce this issue now.

Is there any chance you could include the boost.bc file and the boost headers you are using in the git repo?

**Comment by sbc100** [comment]
I was able to build and run thids example against boost-1.57.0 using `-flto`.   I suspect that it might be related to the fact that you are pre-linking the objects from boost into a single `.bc` file.  

In my build I use `emar` to create libboost.a and link against that.   I will try linking the object together into a single bc.

Could you try using `emar` and see if the issue goes away.

```
$ emcc main.o libfilesystem.a  -s SAFE_HEAP=1 -s NO_EXIT_RUNTIME=1 -s DISABLE_EXCEPTION_CATCHING=0 -s LLD_REPORT_UNDEFINED --profiling -flto
$ $ $NODE main.js 
/temp/dir
```

**Comment by sbc100** [comment]
Yeah this works for me even when I build libboost.bc as a single large bc file.

**Comment by sbc100** [comment]
Here is the Makefile I used in place of your build.bat:

```
BOOST_INCLUDE_PATH=../../boost
SRCS=$(wildcard ../../boost/libs/filesystem/src/*.cpp ../../boost/libs/system/src/*.cpp)
OBJS=$(patsubst %.cpp,%.o,$(SRCS))

all: main.js

clean:
	rm -f main.o main.js libfilesystem.* $(OBJS)

%.o: %.cpp
	emcc $^ -I$(BOOST_INCLUDE_PATH) -std=c++17 -D_LIBCPP_ENABLE_CXX17_REMOVED_AUTO_PTR -s MAIN_MODULE=2 -s SAFE_HEAP=1 -s DISABLE_EXCEPTION_CATCHING=0 -flto -o $@ -c

libfilesystem.a: $(OBJS)
	emar cr $@ $^

libfilesystem.bc: $(OBJS)
	emcc -r -o $@ $^ -flto

main.js: main.o libfilesystem.bc
	EMCC_FORCE_STDLIBS='libc,libc++,libc++abi' emcc $^ $(BOOST_BC_PATH) -s SAFE_HEAP=1 -s NO_EXIT_RUNTIME=1 -s DISABLE_EXCEPTION_CATCHING=0 --profiling -flto -o $@
```

**Comment by spsaxena** [comment]
Thanks @sbc100 for trying out and sorry for delayed response.

The test case I shared was created with minimal settings. So if other than basic settings needed to compile, any of the settings is changed (say remove flto or remove MAIN_MODULE or remove DISABLE_EXCEPTION_CATCHING=0) then I dont hit segmentation fault and the app runs fine.
With either MAIN_MODULE=1 or MAIN_MODULE=2 we encounter segmentation fault.

I can still reproduce the issue, so I am attaching all details.

```
- fetch boost
F:\git3dParty\boost>git checkout boost-1.57.0
So HEAD is now at 3913b6bb7

- setup emscripten
F:\git3dParty\emsdk>emsdk install latest & emsdk activate latest
F:\git3dParty\emsdk>emcc --version
emcc (Emscripten gcc/clang-like replacement) 1.39.10 (commit 1bd7d547598f3fc74699c172f6c9c59a1e8484f1)

(had some issues with bat file so pasting command line as is)

- build boost.bc
F:\OldStuff\Learn\Emscripten\segFault>emcc F:/git3dParty/boost/libs/filesystem/src/codecvt_error_category.cpp F:/git3dParty/boost/libs/filesystem/src/operations.cpp F:/git3dParty/boost/libs/filesystem/src/path.cpp F:/git3dParty/boost/libs/filesystem/src/path_traits.cpp F:/git3dParty/boost/libs/filesystem/src/portability.cpp F:/git3dParty/boost/libs/filesystem/src/unique_path.cpp F:/git3dParty/boost/libs/filesystem/src/utf8_codecvt_facet.cpp F:/git3dParty/boost/libs/system/src/error_code.cpp -IF:/git3dParty/boost -std=c++17 -D_LIBCPP_ENABLE_CXX17_REMOVED_AUTO_PTR -s ERROR_ON_UNDEFINED_SYMBOLS=1 -s WASM_OBJECT_FILES=0 -flto -fPIC -Oz -r -o boost.bc

Minor issue with config submodule
https://github.com/boostorg/config/tree/boost-1.57.0/include/boost/config/workaround.hpp
is missing but referenced from this location, so copied over from 
F:\git3dParty\boost\libs\config\include\boost\detail\workaround.hpp

- command line to build app
set 'EMCC_FORCE_STDLIBS' = 'libc,libc++,libc++abi'
F:\OldStuff\Learn\Emscripten\segFault>emcc main.cpp F:/OldStuff/Learn/Emscripten/segFault/boost.bc -IF:/GIT_OUTPUT_LL/COMPTOP/Fabric_WebReleaseAsmjs32/refs/boost-headers/Inc/boost-1_57 -std=c++17 -D_LIBCPP_ENABLE_CXX17_REMOVED_AUTO_PTR -s MAIN_MODULE=2 -s SAFE_HEAP=1 -s NO_EXIT_RUNTIME=1 -s DISABLE_EXCEPTION_CATCHING=0 --profiling -flto -fPIC

- run
node a.out.js

- result
F:\OldStuff\Learn\Emscripten\segFault>node a.out.js
segmentation fault
segmentation fault
exception thrown: RuntimeError: abort(segmentation fault) at Error
    at jsStackTrace (F:\OldStuff\Learn\Emscripten\segFault\a.out.js:2390:17)
    at stackTrace (F:\OldStuff\Learn\Emscripten\segFault\a.out.js:2407:16)
    at abort (F:\OldStuff\Learn\Emscripten\segFault\a.out.js:2154:44)
    at segfault (F:\OldStuff\Learn\Emscripten\segFault\a.out.js:1169:3)
    at SAFE_HEAP_LOAD_i32_1_U_1 (wasm-function[2123]:30)
    at std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> >::__is_long() const (wasm-function[222]:48)
    at std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> >::size() const (wasm-function[211]:49)
    at std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> >::empty() const (wasm-function[214]:43)
    at boost::filesystem::path::wstring() const (wasm-function[356]:79)
    at main (wasm-function[2075]:226),RuntimeError: abort(segmentation fault) at Error
    at jsStackTrace (F:\OldStuff\Learn\Emscripten\segFault\a.out.js:2390:17)
    at stackTrace (F:\OldStuff\Learn\Emscripten\segFault\a.out.js:2407:16)
    at abort (F:\OldStuff\Learn\Emscripten\segFault\a.out.js:2154:44)
    at segfault (F:\OldStuff\Learn\Emscripten\segFault\a.out.js:1169:3)
    at SAFE_HEAP_LOAD_i32_1_U_1 (wasm-function[2123]:30)
    at std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> >::__is_long() const (wasm-function[222]:48)
    at std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> >::size() const (wasm-function[211]:49)
    at std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> >::empty() const (wasm-function[214]:43)
    at boost::filesystem::path::wstring() const (wasm-function[356]:79)
    at main (wasm-function[2075]:226)
    at abort (F:\OldStuff\Learn\Emscripten\segFault\a.out.js:2160:9)
    at segfault (F:\OldStuff\Learn\Emscripten\segFault\a.out.js:1169:3)
    at SAFE_HEAP_LOAD_i32_1_U_1 (wasm-function[2123]:30)
    at std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> >::__is_long() const (wasm-function[222]:48)
    at std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> >::size() const (wasm-function[211]:49)
    at std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> >::empty() const (wasm-function[214]:43)
    at boost::filesystem::path::wstring() const (wasm-function[356]:79)
    at main (wasm-function[2075]:226)
    at Module._main (F:\OldStuff\Learn\Emscripten\segFault\a.out.js:5748:32)
    at callMain (F:\OldStuff\Learn\Emscripten\segFault\a.out.js:6598:15)
```

I have also attached build including boost.bc  
[segFault.zip](https://github.com/emscripten-core/emscripten/files/4855934/segFault.zip)

**Comment by sbc100** [comment]
I think the problem is with your main.cpp and the tempPath reference.   This is a C++ reference (&) that is constructed with a reference to rvalue temporary.

```static const auto& tempPath(boost::filesystem::path(L"/temp") / L"dir");```

I don't think this code is safe since the tempPath reference is essentially a pointer to a temporary whish is destroyed by the time the print operation happens.

If you simply remove the `&` the problem goes away.

**Comment by sbc100** [comment]
Actually these must be something else going on here since it works without MAIN_MODULE ..

**Comment by spsaxena** [comment]
I dont see other compilers complaining about code. Tried building with c++ (Ubuntu 9.3.0-10ubuntu2), builds and runs without any issue
[segfault.zip](https://github.com/emscripten-core/emscripten/files/4885317/segfault.zip)

**Comment by sbc100** [comment]
I'm managed to create a much smaller reproducer.    I'm having quite a hard time tracking down the precise cause though.

Here is a simple version that shows that same issue I believe, but does not depend on boost at all.

```
#include <stdio.h>
#include <string>
#include <assert.h>

class Foo {
public:
  Foo() { printf("new foo %p\n", this); }
  ~Foo() { printf("del foo %p\n", this); }
  Foo dup() { printf("dup foo %p\n", this); return *this; }
};

static const Foo& tempPath(Foo().dup());

int main() {
  printf("ptr: %p\n", &tempPath);
  assert(&tempPath);
  return 0;
}
```

**Comment by sbc100** [comment]
tempPath is null reference when MAIN_MODULE + lto is used, but valid pointer otherwise.

**Comment by spsaxena** [comment]
Great, wish your code reproduces same issue.
Interestingly if I use tempPath in your code to say make a call, the call is successful. No segmentation fault or crash.
Like I added 
`void call() const { printf("hello\n"); } ` // to class Foo
and called using tempPath
`tempPath.call();`
the call goes thru and hello is printed out. 
With Ubuntu c++ the tempPath is valid reference also.

![image](https://user-images.githubusercontent.com/12669021/86862058-09164480-c0fb-11ea-900a-acff40cfe406.png)

Do you think this could be optimizer issue.

**Comment by sbc100** [comment]
Its something pretty strange because it only happens with LTO.    

It does seem like an odd code pattern though.   Wouldn't it be better not to use a reference here and just declare a local object?   I know we clearly need to fix this, but maybe it could also be fixed by modifying the input code.

Doesn't it seem kind of wrong to declare a global reference to a rvalue?

**Comment by spsaxena** [comment]
Thanks @sbc100 for accepting as a bug.

|Doesn't it seem kind of wrong to declare a global reference to a rvalue?
The code is valid. I find a similar article on [stackoverflow](https://stackoverflow.com/questions/4615491/static-const-references-to-temporaries). I also posted the question to [isocpp forum](https://lists.isocpp.org/std-discussion/current/0622.php) and got the same reply, that the code is valid c++ code with an explanation.  

Though I fixed it at places in my code. But I feel its quite important to have a fix after someone working on other platform introduced a similar code elsewhere which again broke it for web. And being a runtime issue it's difficult to catch it unless we hit that.

**Comment by spsaxena** [comment]
hi @sbc100 did you get a chance to look into it, or if you can comment on when can we expect the fix

**Comment by spsaxena** [comment]
hi @sbc100 any updates on this issue

**Comment by sbc100** [comment]
Sorry, I've not had time to look into it any deeper.   I will try and take another look this week.  If I remember right it seems like a very hard to pin down issue.

**Comment by spsaxena** [comment]
thank you and yes being a runtime issue its caught only when the workflow is executed, so high chance to miss it during testing.

**Comment by spsaxena** [comment]
hi @sbc100 any updates on this issue

**Comment by sbc100** [comment]
So it looks like this issue was fixed by a recent binaryen change:

https://github.com/WebAssembly/binaryen/pull/2953

Great news!

The first emscripten release to contain this change was (I believe) 1.39.20.

Can you confirm that fixes the issue?

**Comment by spsaxena** [comment]
hurrae!! yes the issue is fixed as tested with 1.39.20. 
Thank you, you may close it now.
